File: app/src/main/java/com/edricchan/studybuddy/RegisterActivity.java
package com.edricchan.studybuddy;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.FirebaseAuth;

import androidx.appcompat.app.AppCompatActivity;

public class RegisterActivity extends AppCompatActivity {

	private TextInputLayout inputEmail, inputPassword;
	private MaterialButton btnSignIn, btnSignUp;
	private ProgressBar progressBar;
	private FirebaseAuth auth;
	private static final String TAG = SharedHelper.getTag(RegisterActivity.class);

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_register);
		getSupportActionBar().setDisplayHomeAsUpEnabled(true);
		//Get Firebase auth instance
		auth = FirebaseAuth.getInstance();

		btnSignIn = findViewById(R.id.signInBtn);
		btnSignUp = findViewById(R.id.signUpBtn);
		inputEmail = findViewById(R.id.emailRegister);
		inputPassword = findViewById(R.id.passwordRegister);
		progressBar = findViewById(R.id.progressBar);

		btnSignIn.setOnClickListener(v -> {
			startActivity(new Intent(RegisterActivity.this, LoginActivity.class));
			finish();
		});

		btnSignUp.setOnClickListener(v -> {

			String email = SharedHelper.getEditTextString(inputEmail).trim();
			String password = SharedHelper.getEditTextString(inputPassword).trim();

			if (TextUtils.isEmpty(email)) {
				Snackbar.make(findViewById(R.id.registerActivity), "Please enter an email address", Snackbar.LENGTH_LONG)
						.show();
				return;
			}

			if (TextUtils.isEmpty(password)) {
				Snackbar.make(findViewById(R.id.registerActivity), "Please enter a password!", Snackbar.LENGTH_LONG)
						.show();
				return;
			}

			if (password.length() < 6) {
				Snackbar.make(findViewById(R.id.registerActivity), "A minimum of 6 characters is required for a password", Snackbar.LENGTH_LONG)
						.show();
				return;
			}

			progressBar.setVisibility(View.VISIBLE);
			//create user
			auth.createUserWithEmailAndPassword(email, password)
					.addOnCompleteListener(RegisterActivity.this, task -> {
						progressBar.setVisibility(View.GONE);
						if (task.isSuccessful()) {
							startActivity(new Intent(RegisterActivity.this, MainActivity.class));
							finish();
						} else {
							Snackbar.make(findViewById(R.id.registerActivity), "An error occurred while authenticating. Try again later.", Snackbar.LENGTH_LONG)
									.show();
						}
					});

		});
		checkNetwork();
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case android.R.id.home:
				onBackPressed();
				return true;
			default:
				return super.onOptionsItemSelected(item);
		}
	}

	private void checkNetwork() {
		if (SharedHelper.isNetworkAvailable(this)) {
			btnSignUp.setEnabled(true);
			btnSignIn.setEnabled(true);
			inputEmail.setEnabled(true);
			inputPassword.setEnabled(true);
		} else {
			btnSignUp.setEnabled(false);
			btnSignIn.setEnabled(false);
			inputEmail.setEnabled(false);
			inputPassword.setEnabled(false);
			Snackbar.make(findViewById(R.id.registerActivity), "No internet connection available. Some actions are disabled", Snackbar.LENGTH_INDEFINITE)
					.setBehavior(new NoSwipeBehavior())
					.setAction("Retry", view -> checkNetwork()).show();
		}
	}

	@Override
	protected void onResume() {
		super.onResume();
		progressBar.setVisibility(View.GONE);
	}
}

File: app/src/main/java/com/edricchan/studybuddy/RegisterActivity.kt
package com.edricchan.studybuddy

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.edricchan.studybuddy.extensions.editTextStrValue
import com.edricchan.studybuddy.extensions.isInvalidEmail
import com.edricchan.studybuddy.utils.SharedUtils
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import kotlinx.android.synthetic.main.activity_register.*

class RegisterActivity : AppCompatActivity() {

	private var auth: FirebaseAuth = FirebaseAuth.getInstance()

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_register)
		supportActionBar?.setDisplayHomeAsUpEnabled(true)
		auth = FirebaseAuth.getInstance()

		signInBtn.setOnClickListener {
			startActivity(Intent(this, LoginActivity::class.java))
			finish()
		}

		signUpBtn.setOnClickListener {
			val email = emailTextInputLayout.editTextStrValue
			val password = passwordTextInputLayout.editTextStrValue

			if (email.isNullOrBlank()) {
				if (
						emailTextInputLayout.error.isNullOrEmpty() ||
						emailTextInputLayout.error == getString(R.string.edittext_errors_invalid_email)
				) {
					emailTextInputLayout.error = getString(R.string.edittext_errors_empty_email)
				}
			} else if (email.isInvalidEmail()) {
				if (
						emailTextInputLayout.error.isNullOrEmpty() ||
						emailTextInputLayout.error == getString(R.string.edittext_errors_empty_email)
				) {
					emailTextInputLayout.error = getString(R.string.edittext_errors_invalid_email)
				}
			} else {
				if (emailTextInputLayout.error != null) {
					if (emailTextInputLayout.error!!.isNotEmpty()) {
						emailTextInputLayout.error = null
					}
				}
			}
			if (password.isNullOrEmpty()) {
				if (
						passwordTextInputLayout.error.isNullOrEmpty() ||
						passwordTextInputLayout.error == getString(R.string.edittext_errors_invalid_password)
				) {
					passwordTextInputLayout.error = getString(R.string.edittext_errors_empty_password)
				}
			} else if (password.length < 6) {
				if (
						passwordTextInputLayout.error.isNullOrEmpty() ||
						passwordTextInputLayout.error == getString(R.string.edittext_errors_empty_password)
				) {
					passwordTextInputLayout.error = getString(R.string.edittext_errors_invalid_password)
				}
			} else {
				if (passwordTextInputLayout.error != null) {
					if (passwordTextInputLayout.error!!.isNotEmpty()) {
						passwordTextInputLayout.error = null
					}
				}
			}
			if (passwordTextInputLayout.error!!.isNotEmpty() || emailTextInputLayout.error!!.isNotEmpty()) {
				return@setOnClickListener
			}

			progressBar?.visibility = View.VISIBLE
			// Assume that email and password are non-null
			auth.createUserWithEmailAndPassword(email!!, password!!)
					.addOnCompleteListener(this@RegisterActivity) { task ->
						progressBar?.visibility = View.GONE
						if (task.isSuccessful) {
							startActivity(Intent(this@RegisterActivity, MainActivity::class.java))
							finish()
						} else {
							Snackbar.make(findViewById(R.id.registerActivity), "An error occurred while authenticating. Try again later.", Snackbar.LENGTH_LONG)
									.show()
							Log.e(TAG, "An error occurred while authenticating.", task.exception)
						}
					}

		}
		checkNetwork()
	}

	override fun onResume() {
		super.onResume()
		progressBar?.visibility = View.GONE
	}

	override fun onOptionsItemSelected(item: MenuItem): Boolean {
		return when (item.itemId) {
			android.R.id.home -> {
				onBackPressed()
				true
			}
			else -> super.onOptionsItemSelected(item)
		}
	}

	private fun checkNetwork() {
		if (SharedUtils.isNetworkAvailable(this)) {
			setViewsEnabled(true)
		} else {
			setViewsEnabled(false)
			Snackbar.make(findViewById(R.id.registerActivity), "No internet connection available. Some actions are disabled", Snackbar.LENGTH_INDEFINITE)
					.setBehavior(NoSwipeBehavior())
					.setAction("Retry") { checkNetwork() }.show()
		}
	}

	/**
	 * Sets all views as shown/hidden
	 *
	 * @param enabled Whether to show the views
	 */
	private fun setViewsEnabled(enabled: Boolean) {
		signUpBtn.isEnabled = enabled
		signInBtn.isEnabled = enabled
		emailTextInputLayout.isEnabled = enabled
		passwordTextInputLayout.isEnabled = enabled
	}

	companion object {
		private val TAG = SharedUtils.getTag(this::class.java)
	}
}